### ISSUE TITLE
Properly handle the Promise returned by <audio>.play()

### ISSUE THREAD

**Issue by ericdrobinson** [MoePlayer/APlayer#179]
# Background
In modern browsers, the `HTMLMediaElement.play()` method is not always synchronous. [According to MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play):

>The `HTMLMediaElement.play()` method attempts to begin playback of the media and returns a [`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) which is fulfilled when the playback has been successfully started, and which is rejected if playback fails to begin for any reason (such as permission issues or other problems).

It seems that "permission issues" cover mobile/recent desktop browser decisions to ignore [`autoplay`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/autoplay). Google has published an article about this [here](https://developers.google.com/web/updates/2016/03/play-returns-promise) which points to [this example page](https://googlechrome.github.io/samples/play-return-promise/). Running that example on an iPhone indeed reports a permissions issue in the output.

In fact, the "permissions" issue for `autoplay` is no longer limited to **mobile** contexts. Safari now [disables autoplay functionality on the Desktop](https://webkit.org/blog/7734/auto-play-policy-changes-for-macos/) by default as well. Opening Google's example page in Safari shows this error:

>Attempting to play automatically...
The play() Promise rejected!
Use the Play button instead.
{"code":35,"name":"NotAllowedError","message":"The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission.","line":124,"column":47,"sourceURL":"https://googlechrome.github.io/samples/play-return-promise/"}

Google has also [published a document](https://developers.google.com/web/updates/2017/06/play-request-was-interrupted) that goes into greater detail about the causes of this. Of particular interest is this note:

>**Note:** Calling `video.pause()` isn't the only way to interrupt a video. You can entirely reset the video playback state, including the buffer, with `video.load()` and `video.src = ''`.

(Note that the same applies for `audio`.)

# In APlayer
Currently, APlayer does not attempt to capture the returned Promise. This is the latest from time of writing:
https://github.com/MoePlayer/APlayer/blob/81002c0bf93425884bd565147ad3c2b85762ca44/src/APlayer.js#L738-L745

Proper handling of the `<audio>` element requires handling the Promise. It is exceptionally simple to cause exceptions with the `<audio>` API (see: #155).

In fact, the basic [Demo Page](https://aplayer.js.org) shows this in the console directly after the APlayer version report:

>[Error] Unhandled Promise Rejection: NotAllowedError (DOM Exception 35): The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission.

I have created a [JSFiddle that shows a range of ways to deal with the HTMLMediaElement API](https://jsfiddle.net/xos51sgf/4/). Try each of the four buttons and check the JavaScript console and output block. Please also take a look at the implementation of the `best()` function. It shows a unified way to handle the API in environments that return Promises and `undefined`.

I would highly recommend that the APlayer codebase be updated to properly handle Promises. This may also require holding a reference to the most recent Promise returned by the `HTMLMediaElement.play()` function (as well as a "playPending" boolean?) so that you can chain other calls (`pause()`, `load()`, etc.) off of it.

**Comment by DIYgod** [comment]
Thanks for your detailed suggestion and wonderful demo.

According to [caniuse](https://caniuse.com/#search=Promise), I find `Promise` has some compatibility problem, `best()` function may not be the best way actually, so I prefer use the way of `better()` function in the next version.

**Comment by ericdrobinson** [comment]
>According to caniuse, I find Promise has some compatibility problem

Are minimum browser requirements for APlayer defined somewhere?

[_I only realized that "legacy" IE was a target by finding [bugs](https://github.com/MoePlayer/APlayer/issues/133) about it in the Issues..._]

>`best()` function may not be the best way actually, so I prefer use the way of `better()` function in the next version.

While this is precisely why I included the different options in the demo, I would also point out that you can very easily **polyfill** promises for older browsers (including IE). See:

1. **[promise-polyfill](https://github.com/taylorhakes/promise-polyfill)** - IE/Edge may benefit from an [additional polyfill](https://github.com/taylorhakes/promise-polyfill#performance).
1. **[es6-promise](https://github.com/stefanpenner/es6-promise)**

[For both of the above, IE9 and lower may need [special handling](https://github.com/stefanpenner/es6-promise#usage-in-ie9).]

The benefit here is that once you decide to drop Internet Explorer as a browser, you can simply remove the polyfill altogether. Then everything will simply work with the `best` method automatically and you won't need to refactor anything. I would **highly recommend** investigating this approach.

**Comment by ericdrobinson** [comment]
>Not only IE, some Android browsers may also have some problems, especially in China.

Ahh, I can see that. Particularly the older Android devices.

>Polyfill is too big for this project.

This would be your call, of course :) I should note that promise-polyfill is "[<1KB Gzipped](https://github.com/taylorhakes/promise-polyfill#promise-polyfill)".

>If the `better()` function and the `best()` function have the same effect, the more compatible way is the better way, isn't it?

That depends, I think. Which is more important: reducing technical debt [and maintenance time/effort] or optimizing download size? This would be your call.

Is 1-2KB of size really all that much? APlayer.min.js is [currently 24.6KB](https://github.com/MoePlayer/APlayer/blob/master/dist/APlayer.min.js). Technically, you could also restrict the number of people impacted by that 1-2KB by only downloading it in legacy browsers that need it ([here is a description](https://philipwalton.com/articles/loading-polyfills-only-when-needed/) of how you might achieve this).

I do not know what the impact of adding 1-2KB to APlayer would be for the overall userbase so I don't feel that I could make a call. My previous recommendation to use a polyfill was solely based on the idea that it will be simpler for you and others to maintain going forward to provide the polyfill. :)

**Comment by DIYgod** [comment]
Sorry, I didn't see it, 1KB is acceptable.

In this way, the `best` method is better, I'll use the `best` method, thanks.

**Comment by ericdrobinson** [comment]
No problem! Happy to be of help! Thanks very much for creating and supporting APlayer!

**Comment by DIYgod** [comment]
Done. https://github.com/MoePlayer/APlayer/commit/c9e7f8aa1101da00b1e4042fddd4b58d6a73a305

**Comment by ericdrobinson** [comment]
@DIYgod Thanks for working on this! Unfortunately, the referenced commit (c9e7f8a) gets the ball rolling but does not actually resolve this issue. The JSFiddle example had local Promise handling in order to keep things as tight as possible. The minimal example shows `play()` and `pause()` being called in the same function context. In practice, this would be extremely rare - more likely is that some APlayer user has some logic that call's APlayer's `play()` and `pause()` methods in some sequence of function calls. If you don't call `pause()` in a context that is chained off the `play()` call, then you will see DOMExceptions arise. The `catch` statement in the `AudioPlayer.play()` setup may trap those, but it's not how the underlying system was intended to work and there's no guarantee what state the player will be in when its done (I think). In other words, a call to `pause()` while `play()` is still processing may trigger an exception but result in the call to `pause()` being _ignored_.

Effectively, you want to keep chaining off the initial call, rather than recreating the Promise chain each time you call `APlayer.play()`. (Something like [this](https://stackoverflow.com/a/21372335/5285364), except across function boundaries...) This ends up working even with the `best` method mentioned in a previous comment because you can "insert" Promises into chains by [returning them in the promise chain itself](https://github.com/nzakas/understandinges6/blob/master/manuscript/11-Promises.md#returning-promises-in-promise-chains).

In a player context (like APlayer), you will need to keep track of the most recent Promise in the chain - specifically, one for each Audio element. Then you can wrap calls to `audio.play()`, `audio.pause()`, and `audio.src =` with Promise chaining logic. Therefore, you might want to set things up like this:

In [`APlayer.initAudio()`](https://github.com/MoePlayer/APlayer/blob/c9e7f8aa1101da00b1e4042fddd4b58d6a73a305/src/js/player.js#L132):
```js
// Start the promise chain with a resolved promise. Removes the
//  need to check for _playedPromise being undefined.
this.audio._playedPromise = Promise.resolve();
```
In [`APlayer.play()`](https://github.com/MoePlayer/APlayer/blob/c9e7f8aa1101da00b1e4042fddd4b58d6a73a305/src/js/player.js#L263)
```js
this.audio._playedPromise = this._playedPromise.then(() => {
    return Promise.resolve(this.audio.play());
})
.catch(() => {
    this.pause();
});
```
In [`APlayer.pause()`](https://github.com/MoePlayer/APlayer/blob/c9e7f8aa1101da00b1e4042fddd4b58d6a73a305/src/js/player.js#L294)
```js
this.audio._playedPromise = this._playedPromise.then(() => {
    this.audio.pause();
    this.timer.disable('progress');
})
.catch(()=>{});
```
In [`APlayer.setMusic()`](https://github.com/MoePlayer/APlayer/blob/c9e7f8aa1101da00b1e4042fddd4b58d6a73a305/src/js/player.js#L211)
```js
this.audio._playedPromise = this._playedPromise.then(() => {
    this.audio.currentTime = 0;
    this.audio.src = this.music.url;
})
.catch(()=>{});
```
Note a few things about this:

1. We ensure that a `catch` is defined for each of these. Even if the [`onRejected` callback](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch#Parameters) is a no-op, adding it will ensure that we can chain in the event that a rogue exception or unexpected rejection occurs.
1. The examples above only capture some of the logic that you may want to place within the Promise chain. Specifically, does it make sense to [adjust the boolean `APlayer.paused` flag](https://github.com/MoePlayer/APlayer/blob/c9e7f8aa1101da00b1e4042fddd4b58d6a73a305/src/js/player.js#L296) if the `audio.pause()` call is delayed (i.e. it's "pending" in the Promise chain), while the audio element wraps up some internal work resulting from a previous `play` call? It may make sense to do the external state adjustment stuff as part of the Promise chain as well because you might get into a broken state if something rejects (your state was updated to be "paused" but an exception earlier in the chain skipped over the `then()` that contained the actual `audio.pause()` statement).

    With this in mind, you may want to push just about everything in [`play()`](https://github.com/MoePlayer/APlayer/blob/c9e7f8aa1101da00b1e4042fddd4b58d6a73a305/src/js/player.js#L263) into a `then()` after the [call to `this.audio.play()`](https://github.com/MoePlayer/APlayer/blob/c9e7f8aa1101da00b1e4042fddd4b58d6a73a305/src/js/player.js#L274) and before [the `catch()`](https://github.com/MoePlayer/APlayer/blob/c9e7f8aa1101da00b1e4042fddd4b58d6a73a305/src/js/player.js#L275). That way, if something breaks and the audio element doesn't _actually_ get to play, your state hasn't been prematurely updated to _look_ like it's playing.

    The same would go for `pause()` and `setMusic()`. The difference is that you could probably wrap the contents of those entire functions in the `then` definitions I proposed up above, rather than break things up as with `play()`. This is because the only Audio element API that can return a Promise is `Audio.play()`.

Does this make sense?

**Comment by DIYgod** [comment]
Thanks, I have fix `catch()` and put everything in `play` `pause` `setAudio` as part of the Promise chain. cf5972879a27459445aaf730f7777c9400667881

And in this version, one player has only one audio element, so I have only one Promise chain per player.

**Comment by ericdrobinson** [comment]
@DIYgod _Aaaaalmost_ there! There is still an oddity with how `this.playedPromise` is set from within calls to `play()`. Basically, the promise chain can currently branch, which could lead to potentially unforeseen consequences (potentially calls happening out-of-order if your luck was really bad).

I wrote a whole bunch about how to resolve this last issue in a [comment to the commit you referenced](https://github.com/MoePlayer/APlayer/commit/cf5972879a27459445aaf730f7777c9400667881#r27989819). Please take a look and let me know if you have any questions!

**Comment by egoist** [comment]
btw you don't really need to include the Promise polyfill in this project, you can just leave that to the users. i.e. You can maintain a cleaner and smaller codebase while allowing users to use polyfills themselves for compatibility with old browsers.

**Comment by DIYgod** [comment]
@ericdrobinson You are right, I fix playedPromise overwritten in your first way(https://github.com/MoePlayer/APlayer/commit/cf5972879a27459445aaf730f7777c9400667881#r27989819) in fb4211a0b050a065f1ecc3e5eb37b97f0bd96ffc.

Does this make sense?

**Comment by ericdrobinson** [comment]
>btw you don't really need to include the Promise polyfill in this project, you can just leave that to the users. i.e. You can maintain a cleaner and smaller codebase while allowing users to use polyfills themselves for compatibility with old browsers.

@egoist Good point. It might be best to create a [New Issue](https://github.com/MoePlayer/APlayer/issues/new) about removing the polyfill and adding instructions for users who wish to support legacy IE/Edge to the README.md.

**Comment by ericdrobinson** [comment]
@DIYgod Looks great! Thanks so much for all the work in getting this sorted out!!

**Comment by DIYgod** [comment]
@ericdrobinson You're welcome! And thank you, too.

However, I want to keep polyfill to make this player easier to use, I think most users still need it as promise's compatibility is not good enough.

**Comment by SevenOutman** [comment]
@ericdrobinson Glad to see it's finally resolved. Gonna keep up with it.

**Comment by SevenOutman** [comment]
@ericdrobinson I encountered some new problems with audio.play() Promise.
When network is poor or somehow fetching media source is really slow, the audio.play() Promise stays `pending`. Since player.pause() waits for audio.play() Promise to resolve, pause seems broken in users' view in this situation.

Any ideas?

**Comment by ericdrobinson** [comment]
@SevenOutman Hmm... interesting. Is this something that is simple to reproduce?

I think it would be helpful to get more information before attempting a fix. Specifically:

1. Does this happen in the [`play` call](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play) alone or will it happen when setting [`src=`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/src) as well?
   1. A call to `play` has to first **load** the audio before it can play. Setting the `src` property also [theoretically?] triggers the loading process. It would be useful to narrow this down to be as specific as possible (e.g. to be able to say: "it is the loading that causes the pause").
1. In what environment(s) does this occur? Specific browsers? All browsers?
1. **If you can test this reliably**, does skipping the promise chain here result in the browser throwing [the Exceptions](https://developers.google.com/web/updates/2017/06/play-request-was-interrupted) outlined in the initial post? If so, what Exception is thrown and what's the message?

## Improve Messaging?

In the short term, the correct action may be to adjust the UI for the user. Specifically, when the user hits the play button, then switch the Play icon to an animated **loading spinner** that gets replaced by either the Pause (⏸) or Play (▶️) button when the Promise resolves or rejects. I would also suggest providing an event API to APlayer implementers for when a load error/timeout occurs (likely what is causing this new issue). That way they might be able to provide the user with a helpful message of something like "Couldn't load the file." Instead of silently failing...

## Allow the User to Stop the System

The promise is tied to the state of a specific Audio element. One way to handle this would be to:

1. Call `stop` on the current audio to add the "Stop" call to the promise chain. (This may not strictly be necessary, but it should help in case something keeps the Audio element alive in the background...)
1. **Replace** the Audio Element with a new one and restart the Promise chain.

Note that this is what might be necessary if someone clicks the "Skip" button while a song in a playlist is stuck loading and the user gets impatient and requests playback of another song in the playlist (that theoretically _might_ not have the same loading problem).

You might also attempt to message the situation to the user while allowing interactivity. You could easily add a loading spinner in a location **separate from** the Play/Pause button and then let the user press Play/Pause while _leaving_ the loading spinner until the initial `play` Promise is handled. This would require adding a new state management system to the code; one that distinguishes between **Current** State and **Desired** State. The current state is reflected by the loading spinner (or the actual song loaded), while the UI is the "Desired" state. For 99.9% of the time those two states would match. For the other 0.1% of the time, the UI would effectively act as a "Target" for the **Current** state.

Before any complex implementations occur, however, I'd suggest getting a bit more information on exactly where/when this occurs. I mention "complex" because adding the loading spinner should be a pretty straightforward way to mitigate the issue in the short-term (if not long-term).

**Comment by SevenOutman** [comment]
@ericdrobinson 

## Condition
This occurred when I was writing [vue-aplayer's demo page](https://vue-aplayer.js.org) as testing is out with Chrome. Songs for the two players on top are stored in Google Cloud Storage. For some reasons, connecting Google service is not so stable from China (where I am located), thus the issue occurred. 

## Reproduciton
I think slowing down the connection to media source is likely to reproduce the problem.
Here are steps I reproduces it on [APlayer's demo page](https://aplayer.js.org)
1. Visit the [demo page](https://aplayer.js.org) in Chrome, and open Chrome DevTools.
2. In 'network' tab, add a custom network throttling profile with minimum download speed and great latency. Then apply this profile.
![2018-03-17 11 05 01](https://user-images.githubusercontent.com/8225666/37551139-1aefaa4e-29d5-11e8-9f5d-bd91d57ec8d8.png)
3. Click the second song in the playlist (first one is preloaded before you apply the network profile), and in your network inspector you can see requests are pending.
![2018-03-17 11 20 52](https://user-images.githubusercontent.com/8225666/37551165-82ea9ae6-29d5-11e8-90e1-d2a2515759a7.png)
4. Now player's thumbnail shows pause button. Try clicking it, it won't work while the requests are still pending. The progress bar shows loading, but still not good experience when user clicks on a button and see nothing happens.
![2018-03-17 11 22 55](https://user-images.githubusercontent.com/8225666/37551170-9508d47c-29d5-11e8-93db-98de5c35e5a0.png)

Theoretically this can occur on any browser that supports audio.play() Promise.

## Key points
The tricky part is, when user clicks on pause button, player doesn't know whether the song has successfully started to play(it simply relies on Promise's resolve()). Maybe audio.play() Promise should be handled in some other approach, e.g. when player.pause() is called before it resolves, abandon the Promise directly, and player.play() returns a (custom) rejected Promise.

**Comment by ericdrobinson** [comment]
@SevenOutman First, I'm seeing lots of `403 (Forbidden)` errors when attempting to play certain songs, including all of the playlist songs (with the exception of the first). Only the first two songs will play, currently. Perhaps test out the page in a Private Browsing mode (e.g. Chrome's Incognito Window)?

## Define the Goal
Your issue is specifically a UX related one and is summed up in step 4 of your Reproduction section:

>Now player's thumbnail shows pause button. Try clicking it, it won't work while the requests are still pending. The progress bar shows loading, but still **not good experience if user clicks a button and see nothing happens.**

The goal here is to reduce frustration. Currently, the player _**looks**_ like it's playing but it isn't. It's actually still _loading_. As such, the correct thing to do is to either put the Pause button into a **disabled** state or add a **loading** UI. That will inform the user that the player is currently busy.

That said, there is another potential issue: switching between songs in playlists. In high-latency scenarios, the user may want to jump to a different song if one takes too long to load. Even if the target song would be quick-to-load, the Promise chain would wait on the currently loading one (if I'm reading the source correctly). To properly message _this_ using a UI state adjustment as in the previous paragraph, the entire list would need to be put into a disabled state (to message the disabling of "click response").

That all said, let's take a deeper look at what we're dealing with in the Promise chain to determine a good solution.

## What the Play Promise Means
[According to Google](https://developers.google.com/web/updates/2016/03/play-returns-promise):

>The Promise won’t fulfill until playback has actually started, meaning the code inside the `then()` will not execute until the media is playing.

This is important because it isn't necessarily tied to loading. It simply says "Now successfully in '_playing_' state." Google's article provides the following non-loading example:

>There are contexts in which a web browser may decide to delay the start of playback—for instance, desktop Chrome will not begin playback of a <video> until the tab is visible.

I mention this because it helps answer the question I posed earlier about whether or not this occurs as part of the loading process. It appears to be specific to when playback actually begins.

All of this boils down to:

1. **Promise _Resolves_**: audio is playing.
1. **Promise _Rejects_**: audio did not (or could not) play and a reason is provided.

Thinking this over, there are three options for how to handle this:

1. Update the UI to better inform the user as to what the status is (add a **loading** visual state).
1. Change the promise flow to allow forcing a promise rejection.
1. Add a system to replace the audio element (and thus the promise).

Let's take a look at each of these below:

### Option 1: Explicitly Message the Loading State
If the audio is not playing, it is very likely loading. This is most likely the state that you happened upon in your testing.  The audio element is _trying_ to play music but it hasn't finished loading. The simplest solution would be to disable the pause button or replace it with UI that indicates "Busy Loading..."

What does it mean for the user to press the pause button in this case? Perhaps that once loading finishes, it should _not_ play? This would require allowing the UI to flow without requiring synchronization with the state of the Promise chain - it would require viewing the UI state as the _**desired**_ state and the promise chain as the _**actual**_ state. This would require some extra work to get moving.

### Option 2: Allow Forcing Promise Rejection
If the audio is slowly loading and the user is tired of waiting, then perhaps the correct thing to do is to intentionally cause the exception to occur (but `catch` it) and break the chain. This would enable, for example, immediately jumping between songs in a playlist - it would unfreeze the user from having to wait to _try_ to play other songs. To enable this, the Promise handling would need to be adjusted such that scheduled functions could be made aware of the fact that a break is expected - likely through a property on the player itself that indicates this.

Another naive approach is possible here. This would involve saying "I don't really care about the Promise stuff" and simply doing something equivalent to the following:

```js
// Swallow any exceptions between call to play() and when audio starts playing.
Promise.resolve(this.audio.play()).catch(()=>{});
```

You _could_ catch Exceptions and turn them into an APlayer `error` event, but this might not be strictly necessary. I personally have no idea if this might cause issues for how browsers handle playback state changing now or in the future.

### Option 3: Replace the Audio Element
This option mostly handles the playlist scenario. It would effectively replace the audio element when switching songs - an operation which would also require replacing the Promise Chain. This reduces the problem to handling the situation of a single audio file being an issue.

Because changing the source (`src`) is one of the ways of causing an exception to arise in the Promise chain, there doesn't appear to be a "clean" way to interrupt the asynchronous process started by `play()`. It seems that the assumption for HTML5 is that each audio file in a webpage would have its own audio element. That said, breaking the chain in this way (**Option 2**) may not be such a terrible thing.

## Moving Forward
How this is handled will require input from @DIYgod. Answering the following questions will help determine the implementation:

1. Should the UI always show the **current** player state or the **desired** player state.
1. Does it make sense to add a **disabled** state to the buttons? (e.g. grey-out the button; replace it with loading spinner; ...)
1. In situations of high-latency, should the user be able to switch between songs in a list?

It might also be really helpful to get some insight from someone who has a deeper understanding of what's going on behind the scenes with Audio Playback promises and whether forcing a promise rejection is an okay approach or if it should be avoided. It seems that @beaufortfrancois might be a good person to ask given that he authored the [Google article](https://developers.google.com/web/updates/2017/06/play-request-was-interrupted) describing fixes for the Promise Exception...

**Comment by SevenOutman** [comment]
@ericdrobinson I think when user want to switch songs, he should be guaranteed able to. So I personally prefer Option 2 for Vue-APlayer, and would probably add it into next patch update.

**Comment by li9269391** [comment]
我遇到的是一样的问题吗

IOS 微信中和 safari 中播放报错：NotAllowedError (DOM Exception 35): The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission.
安卓中报错：NotAllowedError: play() can only be initiated by a user gesture.

**Comment by SevenOutman** [comment]
@li9269391 不是，你遇到的是[另一个问题](https://webkit.org/blog/7734/auto-play-policy-changes-for-macos/)

**Comment by li9269391** [comment]
@SevenOutman 谢谢，我并没有设置自动播放，知道原因了，插件本身用了 promise-polyfill 和我的 webpack 配置构建冲突影响

**Comment by beaufortfrancois** [comment]
@ericdrobinson Do you mind giving a quick summary of the root issue? Thread is quite long and I'm happy when I can ;)

**Comment by DIYgod** [comment]
I agree with @SevenOutman , option 2 is also my want.

For better experience and less waiting, interrupting the pending play promise is obviously better than waiting for promise resolving when users don't want to play it anymore.

In this way, when pausing or switching audio, UI should make change immediately, and the pending play promise should be interrupted immediately.

So my answer is that the UI always show the desired player state, and user operation shouldn't be forbidden.

### Implementation

I think we shouldn't do anything, including handling the play promise, just let the play promise interrupted naturally. It unfortunately returns the original implementation.

I don't find this kind of interrupting has obvious harm, but I'm not sure.

So my implementation just like the follows:

```js
play () {
    // ...
    const playedPromise = audio.play();
    if (playedPromise) {
        playedPromise.catch(() => {
            player.pause();
        });
    }
}
pause () {
    // ...
    audio.pause();
}
switchAudio () {
    // ...
    audio.src = '...';
}
```

**Comment by ericdrobinson** [comment]
@beaufortfrancois
>Do you mind giving a quick summary of the root issue?

APlayer is a custom HTML5 music player ([demo](https://vue-aplayer.js.org/)). As such it allows the user to play/pause songs and skip around between tracks in a playlist. The current version of APlayer features a single audio element per player instance. The player instance interacts with the audio element in order to provide audio playback from the browser context: the player instance provides the UI/UX while the audio element provides the underlying playback capabilities.

In recent versions of Chrome (and most other browsers), the audio element's `play()` call began returning Promises. Under certain circumstances, of which you are aware, the Promises can reject and result in errors in the console.

The examples provided in the various Google Developer articles (including the one you wrote) seem to suggest that the "proper" way to pause an audio element is to chain it off the play promise. This is all well and good for simple use cases, but the player case is a bit more complex. Presently, all calls from the player instance in APlayer that interact with the audio element chain off of the play promise. It works fine generally, but we've hit a snag: imagine a playlist where songs come from disparate servers. One server may have high latency and delay the playback of the song. An impatient user may wish to skip to the next song. If everything is promise-based, then the user will have to wait for the slow server to return enough data to play so that the promise may resolve and the chain can continue.

This suggests that the _correct_ thing to do in this case is to tell the audio element to immediately change its source **and break the promise chain** (cause the exception/error that forces it to reject). This _seems_ like a valid approach but there's a bit of hesitation: what if the promise approach was implemented to help protect the audio element from _breaking_? There is very little written about the promise returned and what is going on internally with the audio element's state.

So the questions we'd like an insider's view on are:
Is it possible that the audio element's underlying systems will get into a broken/unusable state? Or is it totally fine to cause the promise to reject in this manner and simply handle the rejection case?

**Comment by ericdrobinson** [comment]
@DIYgod That makes sense. The one thing I would suggest is that you _don't_ call `pause()` in the rejection handler. In the two cases [that we know of] that cause the promise to reject, the audio will either already be paused or start loading another song.

**Comment by SevenOutman** [comment]
@ericdrobinson Agreed.

**Comment by DIYgod** [comment]
@ericdrobinson Actually, `pause()` is in order to handle `NotAllowedError` which caused by forbidden autoplay in some mobile browsers, but you are right, I need to judge `error.name` before `pause()`, like the follows:

```js
play () {
    // ...
    const playedPromise = audio.play();
    if (playedPromise) {
        playedPromise.catch((e) => {
            if (e.name === 'NotAllowedError') {
                player.pause();
            }
        });
    }
}
```

**Comment by ericdrobinson** [comment]
@DIYgod It may be worthwhile to do that for the two "common" (or expected?) Exception names. [Specifically](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/play#Exceptions):

1. **NotAllowedError** - The user agent (browser) or operating system doesn't allow playback of media in the current context or situation. This may happen, for example, if the browser requires the user to explicitly start media playback by clicking a "play" button.
1. **NotSupportedError** - The media source (which may be specified as a [`MediaStream`](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream), [`MediaSource`](https://developer.mozilla.org/en-US/docs/Web/API/MediaSource), [`Blob`](https://developer.mozilla.org/en-US/docs/Web/API/Blob), or [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File), for example) doesn't represent a supported media format.

Both of those seem like legitimate reasons to call `pause()` and ensure that the UI updates to correctly reflect the state of the audio element.

That said, it may be worthwhile separating the UI update logic from the call to `audio.pause()`. This would allow you to call that code separately. Something like:

```js
setUIPaused () {
    // Adjust buttons and other settings to show pause state.
    // ...
}

pause () {
    this.audio.pause();
    this.setUIPaused();
}

play () {
    // ...
    const playedPromise = audio.play();
    if (playedPromise) {
        playedPromise.catch((e) => {
            if (e.name === 'NotAllowedError') {
                // Already "paused" - just set the UI.
                player.setUIPaused();
            }
        });
    }
}
```

**Comment by DIYgod** [comment]
@ericdrobinson Agree, I have improved it in 5c9456b650768d73d50749241b7325f1d7458d53 .
